### PR TITLE
cdc: rename CDC description tables

### PR DESCRIPTION
**Pull request by kbr-**
Commit 968177da049a455edd8b8b6856698c277635f3f1 has changed the schema
of cdc_topology_description and cdc_description tables in the
system_distributed keyspace.

Unfortunately this was a backwards-incompatible change: these tables
would always be created, irrespective of whether or not "experimental"
was enabled. They just wouldn't be populated with experimental=off.

If the user now tries to upgrade Scylla from a version before this change
to a version after this change, it will work as long as CDC is protected
by the experimental flag and the flag is off.

However, if we drop the flag, or if the user turns experimental on,
weird things will happen, such as nodes refusing to start because they
try to populate cdc_topology_description while assuming a different schema
for this table.

The simplest fix for this problem is to rename the tables. This fix must
get merged in before CDC goes out of experimental.
If the user upgrades his cluster from a pre-rename version, he will simply
have two garbage tables that he is free to delete after upgrading.

sstables and digests need to be regenerated for schema_digest_test since
this commit effectively adds new tables to the system_distributed keyspace.
This doesn't result in schema disagreement because the table is
announced to all nodes through the migration manager.

Fixes https://github.com/scylladb/scylla/issues/6537.

### CODE REVIEW

File: test/boost/schema_change_test.cc [hunk 1]
```diff
@@ -670,11 +670,11 @@ future<> test_schema_digest_does_not_change_with_disabled_features(sstring data_
 
 SEASTAR_TEST_CASE(test_schema_digest_does_not_change) {
     std::vector<utils::UUID> expected_digests{
-        utils::UUID("8182496e-4baf-3a07-91e6-caa140388846"),
-        utils::UUID("a65ea746-4d8a-3e5c-8fbf-5f70c14dbcbc"),
-        utils::UUID("a65ea746-4d8a-3e5c-8fbf-5f70c14dbcbc"),
-        utils::UUID("4c138336-4677-3520-8556-4aab007cfedb"),
-        utils::UUID("4c138336-4677-3520-8556-4aab007cfedb"),
+        utils::UUID("2fb5d448-c537-39d1-9384-5166bcdcaa9a"),
```
Comment (haaawk) [inline test/boost/schema_change_test.cc:673]: Would you still need this if you have left old tables? I think the usual way of handling this is to introduce a schema feature and adding the new table only once this feature is enabled. @avikivity probably knows better.

**Comment by kbr-** [comment]
It's explained in the commit message.

>  This doesn't result in schema disagreement because the table is
announced to all nodes through the migration manager.

**Comment by kbr-** [comment]
The same approach was taken when the `cdc_*_description` tables were first introduced.

**Comment by haaawk** [comment]
> It's explained in the commit message.
> 
> > This doesn't result in schema disagreement because the table is
> > announced to all nodes through the migration manager.

Why do you need to update schema_change_test then? You shouldn't have to.

**Comment by kbr-** [comment]
> Why do you need to update schema_change_test then? You shouldn't have to.

If you understand what this test does, you'll understand why it needs to be updated.
Recommended reading: https://github.com/scylladb/scylla/pull/5485/#discussion_r370136734

> The test is there to make sure that if we upgrade a subset of nodes in a cluster, the subset reaches schema agreement with the rest of the nodes. A comment explains it:
> 
> ```
> // We don't want schema digest to change between Scylla versions because that results in a schema disagreement
> // during rolling upgrade.
> ```
> 
> However, the test is a bit too specific, i.e. it sometimes introduces false positives (by "positive" I mean that the test "catches a bug").
> 
> This happens when:
> 
>     1. the upgraded subset does have a different schema, **but**
> 
>     2. the subset also brings the rest of the nodes up-to-date, so nothing bad actually happened.
> 
> 
> The test catches 1., but it doesn't know about 2.
> 
> My particular situation here is that the upgraded subset announces a new distributed table. This indeed introduces a schema change (1.), but that's an `announce_new_column_family`, so other nodes follow up (2.), and we get schema agreement.
> 
> This happens _by design of the test_. The test author (@tgrabiec) predicted that it would!
> 
> ```
>     const bool regenerate = false;
> ```
> 
> and fortunately, @psarna added a comment which explains what the flag is for:
> 
> ```
>     // NOTICE: Regenerating data for this test may be necessary when a system table is added.
>     // This test uses pre-generated sstables and relies on the fact that they are up to date
>     // with the current system schema. If it is not, the schema will be updated, which will cause
>     // new timestamps to appear and schema digests will not match anymore.
> ```
> 
> This `regenerate` flag exists here for us to use it when necessary. I used it correctly.
> 
> There only way to
> 
> > ensure that the test isn't hiding a schema breakage
> 
> would be to design a new test, which does something completely different. That is out of scope of this issue.

**Comment by kbr-** [comment]
I just did a manual rolling upgrade test with 2 nodes from pre-this commit to after-this commit, with experimental enabled.

As expected, the nodes complained that they couldn't retrieve the CDC generation from system tables.

After removing cdc_local sstables the problem was fixed.

**Comment by haaawk** [comment]
> I just did a manual rolling upgrade test with 2 nodes from pre-this commit to after-this commit, with experimental enabled.
> 
> As expected, the nodes complained that they couldn't retrieve the CDC generation from system tables.
> 
> After removing cdc_local sstables the problem was fixed.

@slivne wanted this to work without removing any tables or any other administrative operation. Is there a way we can improve the experience?

**Comment by kbr-** [comment]
This was with experimental enabled before starting the upgrade! Which we don't officially support.
With experimental disabled the user doesn't have to do anything.

I also just did a manual rolling upgrade test with 2 nodes from pre-this commit with experimental disabled to after-this commit with experimental disabled, and then to experimental enabled.

As expected, cdc_generations and cdc_streams were populated while cdc_*description stayed empty.

**Comment by haaawk** [comment]
> > My particular situation here is that the upgraded subset announces a new distributed table. This indeed introduces a schema change (1.), but that's an `announce_new_column_family`, so other nodes follow up (2.), and we get schema agreement.
> > This happens _by design of the test_. The test author (@tgrabiec) predicted that it would!
> > ```
> >     const bool regenerate = false;
> > ```
> > 

My understanding was that this flag should be used to generate sstables for new tests. Not to change the existing tests. 

> > 
> > and fortunately, @psarna added a comment which explains what the flag is for:
> > ```
> >     // NOTICE: Regenerating data for this test may be necessary when a system table is added.
> >     // This test uses pre-generated sstables and relies on the fact that they are up to date
> >     // with the current system schema. If it is not, the schema will be updated, which will cause
> >     // new timestamps to appear and schema digests will not match anymore.
> > ```

I think this comment is a bit misleading. @psarna what did you have in mind? Certainly regenerating sstables whenever a system table is changed isn't a good idea because it defeats the purpose of this test.

> > 
> > 
> > This `regenerate` flag exists here for us to use it when necessary. I used it correctly.

I'm not so sure about that. Again, my understanding was that regenerate should be used to create new tests. Old tests should always keep working. There is a way to do this by guarding the creation of the new table with a feature that ensures that the new table is added only after the whole cluster expects it.

I know that hiding a distributed table is not necessary so maybe the test shouldn't take distributed tables into digest calculations.

**Comment by kbr-** [comment]
The comment is not misleading, it says exactly what it should.

We are not changing local system tables whose schemas are managed locally by each node (and these tables are used to calculate digests in the test).
We are changing global (distributed) tables whose creation is announced through the migration manager, so each node applies the announced mutation, which means they automatically synchronize the schemas of these tables.

> so maybe the test shouldn't take distributed tables into digest calculations.

Well, the test is IMO completely broken: it's both too weak and too strong.

1. it has false positives. For example, it calculates digests according to the contents of `system_schema.tables` (among other things), which are different if a new distributed table is created (a new row is added into `system_schema.tables`). But as explained above, creating a new distributed table synchronizes nodes.

2. it has false negatives: while it does complain if there's a new distributed table as explained above, it doesn't complain if an existing distributed table changes. Otherwise it would catch the breakage in https://github.com/scylladb/scylla/commit/968177da049a455edd8b8b6856698c277635f3f1 :)

There should be a test  which would catch 2. but not 1.

**Comment by haaawk** [comment]
It is misleading because it says 'when system table is added' and it should say 'when distributed system table is added'. If a local system table is added then we should NEVER regenerate the test and currently, the comment encourages that.

**Comment by kbr-** [comment]
I wonder why the test didn't catch that `system_distributed.cdc_topology_description`'s schema changed.

The test uses `calculate_schema_digest`, which takes into account `system_schema.columns`. This table contains a row says what type does the column `description` of `system_distributed.cdc_topology_description` have. This entry was changed so the test should give another digest.

**Comment by psarna** [comment]
I somehow missed the discussion yesterday - the idea was to simply provide some context for people who see this test for the first time. I intentionally used the words "**may** be necessary when a system table is added", and it's not confined to distributed system tables - for instance, if somebody introduces a new schema table which takes part in the schema digest calculation, the test needs to be regenerated as well in order to provide new cases.

That said, I'm absolutely fine with somebody else rephrasing/extending the comment if it's in any way unclear - it's clear to me, but I'm the author, so my opinion doesn't count (: As long as it helps people grasp what this test case is about, I'm all for it.

**Comment by haaawk** [comment]
@psarna are you sure the addition of a new local schema table should justify the regeneration of the test? In my opinion, such addition should be done in such a way that a new table is not included into digest until whole cluster migrates to the version that includes this new table. Otherwise, rolling upgrade will be broken because there will be schema disagreement until the last old node is upgraded, no? Without regeneration, the test will start failing and that's good because it will catch the problem. If we regenerate it then we will silently hide the error.

**Comment by psarna** [comment]
In the case you presented, we should use `regenerate` to create a new case, guarded by a schema feature, while all previous checksums remain the same. Just like here: 1c0ef5f9e9 - I wrote this comment about `regenerate` when preparing this patch, and after I finally understood what `regenerate` is for.

**Comment by kbr-** [comment]
Tomek has explained the initial purpose of this test in https://github.com/scylladb/scylla/issues/6582.

> The test was supposed to catch accidental changes of the hash, not caused by changing the schema. If you change the schema, the digest is expected to change, and you need to re-generate the test data. That's by design.
(...)
To test that a rolling upgrade works (digest converges) in case the system schema changes, we need to actually run two different versions of scylla, and that cannot be done in a unit test.

In other words, the test is supposed to test only the digest itself, not the entire schema. In particular, the test is supposed to catch only rolling upgrade problems caused by bugs in digest calculation.

**Comment by haaawk** [comment]
Good catch @kbr-. It does explain a lot. I'm afraid that since then the test has evolved and now it checks the schema. Otherwise, we wouldn't have so many test cases that use different schema_feaures.

**Comment by kbr-** [comment]
@psarna maybe you could merge this in?

Also a note on backporting: we don't necessarily need to backport this change.

If we won't backport this change, people simply won't be able to use CDC on clusters upgraded from 4.0 to 4.1, for example. But it's experimental so...

After they upgrade to 4.2, things will work (assuming that 4.2 will include this PR).

**Comment by psarna** [comment]
I will, it's better to merge this one as soon as possible. I'll just double-check that the schema change test works.

**Comment by haaawk** [comment]
It's definitely better to backport this one. Otherwise, 4.1 will be broken.

**Comment by psarna** [comment]
d89b7a0548a9ec60efc5bf456de2396f69341856

**Comment by slivne** [comment]
No backport for 4.0 we can't change that in a release that went out

**Comment by nyh** [comment]
I'll review and commit (if I don't find any obvious problems) now. Would have been nice if previous reviewers could have left a more useful decision ("approve" or "request changes") beyond just discussion.

File: db/system_distributed_keyspace.hh [hunk 2]
```diff
@@ -48,10 +48,10 @@ public:
     static constexpr auto VIEW_BUILD_STATUS = "view_build_status";
 
     /* Nodes use this table to communicate new CDC stream generations to other nodes. */
-    static constexpr auto CDC_TOPOLOGY_DESCRIPTION = "cdc_topology_description";
+    static constexpr auto CDC_TOPOLOGY_DESCRIPTION = "cdc_generations";
```
Comment (nyh) [inline db/system_distributed_keyspace.hh:51]: I would have changed the name of the variable as well. No reason to leave confusing riddles for developers a year from now that have to figure out why the name of the table and the variable is different.
But seeing @psarna 's comment that we should commit this quickly, I'll commit this as-is but please consider sending a followup patch to clear up the confusing names.

File: test/boost/schema_change_test.cc (hunk 1 above)
Comment (nyh) [reply test/boost/schema_change_test.cc:673]: @haaawk I see you "approved" this patch so I assume you got a satisfactory answer for this? Myself, I don't know the answer, and I really dislike and don't understand this test in general and I think we have been doing it the wrong way, but it's not new to this patch :-( I assume that @kbr- ran this test after his change an it passed.

**Comment by nyh** [comment]
Yikes, I see @psarna already committed this patch but forgot to close the PR. closing it now.

**Comment by kbr-** [comment]
@slivne 

can we then backport this to 4.1?
The corresponding dtest change will also have to be backported: https://github.com/scylladb/scylla-dtest/pull/1541 (I assume there's a 4.1 dtest branch).

**Comment by slivne** [comment]
@nyh / @avikivity need a backport to 4.1

**Comment by slivne** [comment]
(there is also a dtest that needs backporting once this is done)

**Comment by haaawk** [comment]
@avikivity / @nyh was this backported?

**Comment by haaawk** [comment]
Apparently not :(

**Comment by avikivity** [comment]
I didn't backport it. I'm leaving it to @nyh.

**Comment by nyh** [comment]
I'll do it now.

**Comment by haaawk** [comment]
Thank you.

**Comment by nyh** [comment]
@avikivity @scylladb/www I see in branch "next-4.1",
```
commit 81dc8eeec74e29ed53b8ef708d67027f32aed847
Author: Kamil Braun <kbraun@scylladb.com>
Date:   Wed Jun 3 12:17:04 2020 +0200
Commit: Avi Kivity <avi@scylladb.com>

    cdc: rename CDC description tables
...
    (cherry picked from commit d89b7a0548a9ec60efc5bf456de2396f69341856)
```
Isn't this the backport you wanted, and @avikivity already done?

**Comment by haaawk** [comment]
oh ok. I looked at 4.1.rc3 and haven't found this there. Sorry if I was wrong.

**Comment by nyh** [comment]
I verified it reached branch-4.1 in June 14th:
```
commit 81dc8eeec74e29ed53b8ef708d67027f32aed847
Author:     Kamil Braun <kbraun@scylladb.com>
AuthorDate: Wed Jun 3 12:17:04 2020 +0200
Commit:     Avi Kivity <avi@scylladb.com>
CommitDate: Sun Jun 14 09:15:36 2020 +0300
```
I don't know when exactly "4.1.rc3" went out. 
I'm removing the backport tags - I think we're done with it.

**Comment by haaawk** [comment]
Thanks @nyh

**Pull request closed without merging**